### ISSUE TITLE
Polygon to mask generation failure?

### ISSUE THREAD

**Issue by horsto** [ultralytics/ultralytics#20142]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

![Image](https://github.com/user-attachments/assets/896e1ebe-4560-4073-974d-fe844e70fcfa)

I am running into mask generation issues (see screenshots abvove). During training data generation, weird artifacts sometimes occur (see triangle "shadows", and weird connections in the mask data). I have also explained this issue on image.sc here: https://forum.image.sc/t/polygon-generation-artifacts/111267 

Has anybody seen this before and knows how / where it is caused? 

### Environment

OS                  macOS-15.3.2-arm64-arm-64bit
Environment         Darwin
Python              3.11.11
Install             pip
Path                /Users/horst/miniconda3/envs/octron/lib/python3.11/site-packages/ultralytics
RAM                 96.00 GB
Disk                985.7/1858.2 GB
CPU                 Apple M2 Max
CPU count           12
GPU                 None
GPU count           None
CUDA                None

numpy               ✅ 2.1.1<=2.1.1,>=1.23.0
matplotlib          ✅ 3.10.1>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.1.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.2>=1.4.1
torch               ✅ 2.6.0>=1.8.0
torch               ✅ 2.6.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.21.0>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0
Out[3]: 
{'OS': 'macOS-15.3.2-arm64-arm-64bit',
 'Environment': 'Darwin',
 'Python': '3.11.11',
 'Install': 'pip',
 'Path': '/Users/horst/miniconda3/envs/octron/lib/python3.11/site-packages/ultralytics',
 'RAM': '96.00 GB',
 'Disk': '985.7/1858.2 GB',
 'CPU': 'Apple M2 Max',
 'CPU count': 12,
 'GPU': None,
 'GPU count': None,
 'CUDA': None,
 'Package Info': {'numpy': '✅ 2.1.1<=2.1.1,>=1.23.0',
  'matplotlib': '✅ 3.10.1>=3.3.0',
  'opencv-python': '✅ 4.11.0.86>=4.6.0',
  'pillow': '✅ 11.1.0>=7.1.2',
  'pyyaml': '✅ 6.0.2>=5.3.1',
  'requests': '✅ 2.32.3>=2.23.0',
  'scipy': '✅ 1.15.2>=1.4.1',
  'torch': '✅ 2.6.0!=2.4.0,>=1.8.0; sys_platform == "win32"',
  'torchvision': '✅ 0.21.0>=0.9.0',
  'tqdm': '✅ 4.67.1>=4.64.0',
  'psutil': '✅ 7.0.0',
  'py-cpuinfo': '✅ 9.0.0',
  'pandas': '✅ 2.2.3>=1.1.4',
  'seaborn': '✅ 0.13.2>=0.11.0',
  'ultralytics-thop': '✅ 2.0.14>=2.0.0'}}

### Minimal Reproducible Example

None

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by Y-T-G** [comment]
This seems more like a labelling issue. The jumps are probably due to polygon points jumping to that place.

**Comment by Y-T-G** [comment]
The polygons are drawn by connecting the consecutive coordinates. So if the next coordinate for some reason moves to some other part of the mask, it would appear like that.

**Comment by horsto** [comment]
Yes, ... but the point is: How do I solve that during polygon generation? 
I am using [imantics](https://imantics.readthedocs.io/en/latest/index.html) polygon generation which itself is based on open cv functions 

```python
  def polygons(self):
        """
        Returns or generates :class:`Polygons` representation of mask.

        :returns: Polygons representation
        :rtype: :class:`Polygons`
        """
        if not self._c_polygons:

            # Generate polygons from mask
            mask = self.array.astype(np.uint8)
            mask = cv2.copyMakeBorder(mask, 1, 1, 1, 1, cv2.BORDER_CONSTANT, value=0)
            polygons = cv2.findContours(mask, cv2.RETR_LIST, cv2.CHAIN_APPROX_SIMPLE, offset=(-1, -1))
            polygons = polygons[0] if len(polygons) == 2 else polygons[1]
            polygons = [polygon.flatten() for polygon in polygons]

            self._c_polygons = Polygons(polygons)
            self._c_polygons._c_mask = self

        return self._c_polygons
```

Any ideas?

**Comment by Y-T-G** [comment]
You can generate polygons from masks using Ultralytics `masks2segments`

https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.masks2segments

**Comment by horsto** [comment]
Ok, seems like the only difference is in the open cv hierarchy of the contours. Ultralytics `masks2segments` uses 
`c = cv2.findContours(x, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)[0]` 

Seems like I might have to play with the hierarchy options in opencv (`RETR_LIST`, `RETR_CCOMP`, `RETR_TREE`).

**Comment by horsto** [comment]
I had a bug in my code. Applying `merge_multi_segment` for fusing lists together solved it. You can find more details about this in this imagesc post: https://forum.image.sc/t/polygon-generation-artifacts/111267
https://github.com/ultralytics/ultralytics/blob/f8870a15b3f4c09d3894bf0266f23d4a96d86418/ultralytics/data/converter.py#L534

**Comment by glenn-jocher** [comment]
Thanks for sharing your solution @horsto! It's great to see you resolved the issue with `merge_multi_segment`. This function connects multiple segmentation pieces by finding minimum distances between segments, which prevents those "jumps" in your polygons that were causing the artifacts. The difference between OpenCV's retrieval modes (`RETR_LIST` vs `RETR_EXTERNAL`) along with proper segment merging is often key to clean mask generation. Your solution will definitely help others who encounter similar issues with polygon generation.
